### PR TITLE
[WIP] Check more type size of push constant

### LINKED ISSUE #2148: Push constant validation issues
We hit a validation error in ANGLE related to the following VVL change de8975779. Oddly enough, this validation error manifests on Nvidia driver 440.10 on Ubuntu 18.04 with a GTX 1660, but not on a more recent driver on Ubuntu 20.04 with a GTX 970.

That said, I believe there are two potential issues here.

First, the push constant resulting in error is:

```
layout(push_constant) uniform PushConstants {
    // Number of levels to generate mipmaps for.
    uint levelCount;
    // Inverse extents of src image for uv calculation.
    vec2 invSrcExtent;
} params;
```

Note that there's an (accidental) implicit padding between the two variables. The validation error is:

```
DebugUtilsMessenger(338): [ UNASSIGNED-CoreValidation-Shader-PushConstantOutOfRange ] Validation Error: [ UNASSIGNED-CoreValidation-Shader-PushConstantOutOfRange ] Object 0: handle = 0x1ad9dfa50050, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x650ba96a | Shader push-constant buffer member 1 at offset 8 is not declared in pipeline layout
                            Object: 0x1ad9dfa50050 (type = Device(3))
```

However, the pipeline correctly does declare 16 bytes of push constants. This seems like a VVL bug.

---

Additionally, I tried to add some debug output to the code that generates this validation error, and I noticed that in `ValidatePushConstantBlockAgainstPipeline`, where the validation happens:

```
if ((range.offset <= offset) && ((range.offset + range.size) >= (offset + size)) &&
    (range.stageFlags & stage)) {
```

for the vec2 field, the calculated size is `6` but should be `8`! That's not causing a failure, but it's definitely a bug.

### PR DESCRIPTION
**Pull request by locke-lunarg**
Calculated the size of multi-dim array, struct, and buffer in push constant, and fixed some bug.  

Fixed #2148
@ShabbyX Could you please test this branch for your bug? Thanks.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -1188,6 +1188,15 @@ bool CoreChecks::ValidateCmdBufDrawState(const CMD_BUFFER_STATE *cb_node, CMD_TY
         } else {
             uint32_t index = 0;
             for (const auto &used : pPipe->push_constant_used_in_shader) {
+                if (index >= cb_node->push_constant_data_set.size()) {
```
Comment (locke-lunarg) [inline layers/core_validation.cpp:1191]: Added a validation if pipeline layout's push constant range is less than shader's push constant range.

File: layers/shader_validation.cpp [hunk 2]
```diff
@@ -1079,62 +1079,81 @@ std::vector<std::pair<descriptor_slot_t, interface_var>> CollectInterfaceByDescr
...
-                        }
+                if (def_member.opcode() == spv::OpTypeStruct) {
+                    SetStructMemberPushConstant(src, def_member, push_constant_used_in_shader, offset, size);
 
-                        if (def_member.opcode() == spv::OpTypeMatrix) {
-                            size *= def_member.word(3);  // matrix's columns. matrix's row is vector.
-                            def_member = src.get_def(def_member.word(2));
-                        }
+                } else if (def_member.opcode() == spv::OpTypePointer) {
+                    // If the member is a buffer, the type will be TpyePointer and struct
```
Comment (ShabbyX) [inline layers/shader_validation.cpp:1110]: typo: TpyePointer

File: layers/shader_validation.cpp [hunk 3]
```diff
@@ -1556,45 +1576,60 @@ bool CoreChecks::ValidatePushConstantBlockAgainstPipeline(std::vector<VkPushCons
...
-                    def_member = src->get_def(def_member.word(2));
-                }
+                if (def_member.opcode() == spv::OpTypeStruct) {
+                    ValidatePushConstantBlockAgainstPipeline(push_constant_ranges, src, def_member, stage, offset, size);
 
-                if (def_member.opcode() == spv::OpTypeVector) {
-                    size *= def_member.word(3);  // vector length
-                }
+                } else if (def_member.opcode() == spv::OpTypePointer) {
+                    // If the member is a buffer, the type will be TpyePointer and struct
```
Comment (ShabbyX) [inline layers/shader_validation.cpp:1594]: Same here

**Review by ShabbyX — COMMENTED** [review]
Sorry, but the failure is only on a trybot that I'm not sure I could trigger a run on. If you were able to trigger bugs based on the push constants declaration in the issue, it should be good. I made a change to remove the accidental padding, so the bot may not trigger the error anymore, but if it still does, we can see what happens after this fix lands.

File: layers/shader_validation.cpp (hunk 2 above)
Comment (locke-lunarg) [reply layers/shader_validation.cpp]: fixed

File: layers/shader_validation.cpp (hunk 3 above)
Comment (locke-lunarg) [reply layers/shader_validation.cpp]: fixed

**Review by karl-lunarg — APPROVED** [review]
I didn't detect any regressions in my testing.  I didn't review the code, so other approvals are necessary.

**Comment by locke-lunarg** [comment]
Closed. Because of a lot of change, I will push a new PR for that.

**Pull request closed without merging**